autoextract_poet/page_inputs.py

```diff
@@ -38,34 +38,44 @@ class _AutoExtractData(Generic[T]):
 
     https://doc.scrapinghub.com/autoextract.html#responses
     """
-
-    item_key: ClassVar[str]
-
+    page_type: ClassVar[str]
     data: dict
+    _item: Optional[T] = attr.ib(init=False, repr=False)
+
+    def __attrs_post_init__(self):
+        # Caching item to avoid building it again and again
+        self._item = self.item_class.from_dict(self.data[self.page_type])
 
     @property
     def item_class(self):
         return self.__orig_bases__[0].__args__[0]
 
     def to_item(self) -> Optional[T]:
-        return self.item_class.from_dict(self.data[self.item_key])
+        return self._item
 
 
 @attr.s(auto_attribs=True)
-class AutoExtractArticleData(_AutoExtractData[Article]):
+class AutoExtractArticleData(AutoExtractData[Article]):
     """Container for AutoExtract Article data.
 
     https://doc.scrapinghub.com/autoextract/article.html
     """
-
-    item_key = "article"
+    page_type = "article"
 
 
 @attr.s(auto_attribs=True)
-class AutoExtractProductData(_AutoExtractData[Product]):
+class AutoExtractProductData(AutoExtractData[Product]):
     """Container for AutoExtract Product data.
 
     https://doc.scrapinghub.com/autoextract/product.html
     """
+    page_type = "product"
 
-    item_key = "product"
```

Comment:
How would you implement container for e.g. WebPage data? See example response at https://doc.scrapinghub.com/autoextract.html#responses. I think that was the reason this field is not called page_type - pageType is an input API argument, and it is not always corresponding to the output fields.

Reply:
> How would you implement container for e.g. WebPage data?

There are several ways of doing that:
* If you want it to be part of the item itself (e.g. product.web_page), then you'll have to add it to the `Product` item and the rest of items and just plug it properly in the `AutoExtractData.__attrs_post_init__()` with the caching approach or in `AutoExtractData.to_item()` otherwise. 
* An alternative is to create a super Item (e.g. ProductResponse) that contains the rest of element `web_page`, `algorithm_version` in addition to the `product` item.
* A better approach than the former is include new methods in `AutoExtractData` that returns this information (e.g. `get_web_page_data()`. 
* Finally, we could follow the approach for `AutoExtractHtml`. If can declare that you need `WebData` and then you get it. 

In definitive, there are many ways of doing it and I don't see anything in the approach in the PR that could block that. 

>  I think that was the reason this field is not called page_type - pageType is an input API argument, and it is not always corresponding to the output fields.

What do you mean? I think this is actually the case, you always get the resultant item under the key that corresponds with `pageType` input. So `pageType` is very precise in this case. I think this will be True even for a multi-pageType case in the future.